File: src/ios/Agent/Chat/MultiPathPlanner.swift
import Foundation

// MARK: - MultiPathPlanner (C12)
// [T-deep-mode-cognitive-p2-c12] Phase 4.5 P2: Multi-path parallel thinking.
//
// When deep mode is on and the model is in the planning phase, the
// deepModeFragment instructs it to generate multiple candidate paths
// (default 3) inside a fenced ```plan``` block, each marked with a
// `## PATH N:` header. The client then picks the best path based on
// a scoring rubric the model itself provides.
//
// TOTAL-SWITCH SAFE: MultiPathPlanner is only invoked from PlanGate
// detection logic, which is itself gated on `deepModeEnabled`. When
// the master switch is off, deepModeFragment is never injected, the
// model never produces multi-path plans, and MultiPathPlanner is
// never called. Zero runtime state, zero persistence, zero UI.

/// A single candidate path in a multi-path plan.
struct CandidatePath: Identifiable, Equatable {
    let id = UUID()
    let index: Int          // 1-based path number
    let title: String       // Short title after "## PATH N:"
    let body: String        // Full plan text for this path
    let rationale: String   // Why this path, from the model's scoring
    let riskLevel: RiskLevel

    enum RiskLevel: String, Equatable {
        case low
        case medium
        case high
    }
}

/// Result of multi-path plan parsing.
enum MultiPathResult: Equatable {
    /// The plan contains multiple candidate paths.
    case multiPath(paths: [CandidatePath], recommendedIndex: Int)
    /// The plan is a standard single-path plan — not multi-path format.
    case singlePath
}

enum MultiPathPlanner {

    /// Header marker for each candidate path in a multi-path plan.
    /// The model is instructed to format each path as:
    /// ```
    /// ## PATH 1: <title>
    /// <steps>
    /// RISK: low|medium|high
    /// ```
    static let pathMarker = "## PATH"
    static let riskMarker = "RISK:"
    static let recommendMarker = "RECOMMENDED:"

    /// Parse a plan text to determine if it's a multi-path plan.
    /// Returns `.singlePath` if the plan doesn't contain multiple
    /// `## PATH` headers (backward compatible with existing PlanGate).
    static func parse(_ planText: String) -> MultiPathResult {
        // Count occurrences of the path marker.
        let lines = planText.components(separatedBy: "\n")
        let pathLines = lines.filter { $0.trimmingCharacters(in: .whitespaces).hasPrefix(pathMarker) }

        guard pathLines.count >= 2 else {
            // Fewer than 2 paths → not a multi-path plan.
            return .singlePath
        }

        // Parse each path block.
        var paths: [CandidatePath] = []
        var recommendedIndex = 1

        var currentPathIndex = 0
        var currentTitle = ""
        var currentBody: [String] = []
        var currentRisk: CandidatePath.RiskLevel = .medium
        var currentRationale = ""

        for line in lines {
            let trimmed = line.trimmingCharacters(in: .whitespaces)

            if trimmed.hasPrefix(pathMarker) {
                // Save previous path if exists.
                if currentPathIndex > 0 {
                    paths.append(CandidatePath(
                        index: currentPathIndex,
                        title: currentTitle,
                        body: currentBody.joined(separator: "\n").trimmingCharacters(in: .whitespacesAndNewlines),
                        rationale: currentRationale.trimmingCharacters(in: .whitespacesAndNewlines),
                        riskLevel: currentRisk
                    ))
                }

                // Start new path.
                currentPathIndex += 1
                // Extract title: "## PATH 1: <title>" → "<title>"
                let afterMarker = String(trimmed.dropFirst(pathMarker.count))
                    .trimmingCharacters(in: .whitespaces)
                // Remove leading number and colon: "1: " or "1 - "
                if let colonRange = afterMarker.range(of: ":") {
                    currentTitle = String(afterMarker[afterMarker.index(after: colonRange.lowerBound)...])
                        .trimmingCharacters(in: .whitespaces)
                } else if let dashRange = afterMarker.range(of: "-") {
                    currentTitle = String(afterMarker[afterMarker.index(after: dashRange.lowerBound)...])
                        .trimmingCharacters(in: .whitespaces)
                } else {
                    currentTitle = afterMarker
                }
                currentBody = []
                currentRisk = .medium
                currentRationale = ""
            } else if trimmed.hasPrefix(riskMarker) {
                let riskStr = String(trimmed.dropFirst(riskMarker.count))
                    .trimmingCharacters(in: .whitespaces)
                    .lowercased()
                currentRisk = CandidatePath.RiskLevel(rawValue: riskStr) ?? .medium
            } else if trimmed.hasPrefix(recommendMarker) {
                // "RECOMMENDED: PATH 2" or "RECOMMENDED: 2"
                let afterMarker = String(trimmed.dropFirst(recommendMarker.count))
                    .trimmingCharacters(in: .whitespaces)
                    .lowercased()
                // Try to extract a number.
                let digits = afterMarker.compactMap { $0.isNumber ? $0 : nil }
                let numStr = String(digits)
                if let num = Int(numStr), num >= 1 && num <= pathLines.count {
                    recommendedIndex = num
                }
            } else {
                // Regular line — add to current path body.
                if currentPathIndex > 0 {
                    currentBody.append(line)
                }
            }
        }

        // Save last path.
        if currentPathIndex > 0 {
            paths.append(CandidatePath(
                index: currentPathIndex,
                title: currentTitle,
                body: currentBody.joined(separator: "\n").trimmingCharacters(in: .whitespacesAndNewlines),
                rationale: currentRationale.trimmingCharacters(in: .whitespacesAndNewlines),
                riskLevel: currentRisk
            ))
        }

        guard paths.count >= 2 else {
            return .singlePath
        }

        return .multiPath(paths: paths, recommendedIndex: min(max(recommendedIndex, 1), paths.count))
    }

    /// Extract the recommended path's step list as a standard plan text,
    /// so it can be fed to WorkflowPlanParser.parseSteps() unchanged.
    static func extractRecommendedPlan(from result: MultiPathResult) -> String? {
        switch result {
        case .singlePath:
            return nil
        case .multiPath(let paths, let recommendedIndex):
            guard paths.indices.contains(recommendedIndex - 1) else { return nil }
            return paths[recommendedIndex - 1].body
        }
    }
}

File: src/android/app/src/main/java/com/openminis/app/agent/MultiPathPlanner.kt
package com.openminis.app.agent

/**
 * Multi-path planner (C12) — 深度龙虾Ai deep mode.
 *
 * When deep mode is on and the model is in the planning phase, the
 * deep mode fragment instructs it to generate multiple candidate paths
 * inside a fenced ```plan``` block, each marked with a `## PATH N:` header,
 * a `RISK:` line and an optional `RECOMMENDED:` line. The client parses these
 * so the UI can offer a Trae-style multi-select of the candidate paths.
 *
 * TOTAL-SWITCH SAFE: only invoked from the plan-gate detection logic, which
 * is itself gated on `deepModeEnabled`. When the master switch is off the
 * model never produces multi-path plans and this object is never called.
 * Zero runtime state, zero persistence.
 */
object MultiPathPlanner {

    /** Header marker for each candidate path. */
    const val PATH_MARKER = "## PATH"

    /** Risk marker within a path block. */
    const val RISK_MARKER = "RISK:"

    /** Recommended-path marker. */
    const val RECOMMEND_MARKER = "RECOMMENDED:"

    /** A single candidate path. */
    data class CandidatePath(
        val index: Int,             // 1-based path number
        val title: String,          // Short title after "## PATH N:"
        val body: String,           // Full plan text for this path
        val rationale: String,      // Why this path, from the model's scoring
        val riskLevel: RiskLevel,
    )

    enum class RiskLevel { LOW, MEDIUM, HIGH }

    sealed class Result {
        /** The plan contains multiple candidate paths. */
        data class MultiPath(val paths: List<CandidatePath>, val recommendedIndex: Int) : Result()

        /** The plan is a standard single-path plan — not multi-path format. */
        object SinglePath : Result()
    }

    /**
     * Parse a plan text to determine if it's a multi-path plan.
     * Returns [Result.SinglePath] if the plan doesn't contain two or more
     * `## PATH` headers (backward compatible with the existing PlanGate).
     */
    fun parse(planText: String): Result {
        val lines = planText.lines()
        val pathLineIndexes = lines.mapIndexedNotNull { i, l ->
            if (l.trimStart().startsWith(PATH_MARKER)) i else null
        }

        if (pathLineIndexes.size < 2) return Result.SinglePath

        val paths = mutableListOf<CandidatePath>()
        var recommendedIndex = 1

        for ((idx, start) in pathLineIndexes.withIndex()) {
            val end = if (idx + 1 < pathLineIndexes.size) pathLineIndexes[idx + 1] else lines.size
            val block = lines.subList(start, end)
            val headerTrimmed = block[0].trim()
            // "## PATH 1: <title>" → "<title>"
            val afterMarker = headerTrimmed
                .removePrefix(PATH_MARKER)
                .trim()
                .substringAfter(":")
                .trim()
            val title = if (afterMarker.isBlank()) headerTrimmed else afterMarker

            var body = ""
            var rationale = ""
            var risk = RiskLevel.MEDIUM
            val bodyLines = mutableListOf<String>()

            for (line in block.drop(1)) {
                val t = line.trim()
                when {
                    t.startsWith(RISK_MARKER) -> {
                        risk = when (t.removePrefix(RISK_MARKER).trim().lowercase()) {
                            "low" -> RiskLevel.LOW
                            "high" -> RiskLevel.HIGH
                            else -> RiskLevel.MEDIUM
                        }
                    }
                    t.startsWith(RECOMMEND_MARKER) -> {
                        val digits = t.removePrefix(RECOMMEND_MARKER)
                            .filter { it.isDigit() }
                        val num = digits.toIntOrNull()
                        if (num != null && num in 1..pathLineIndexes.size) {
                            recommendedIndex = num
                        }
                    }
                    else -> bodyLines.add(line)
                }
            }

            body = bodyLines.joinToString("\n").trim()
            paths.add(
                CandidatePath(
                    index = idx + 1,
                    title = title,
                    body = body,
                    rationale = rationale.trim(),
                    riskLevel = risk,
                )
            )
        }

        if (paths.size < 2) return Result.SinglePath
        return Result.MultiPath(
            paths = paths,
            recommendedIndex = recommendedIndex.coerceIn(1, paths.size),
        )
    }

    /** Extract the recommended path's body for step parsing. */
    fun extractRecommendedPlanBody(result: Result, originalPlan: String): String? {
        return when (result) {
            is Result.SinglePath -> null
            is Result.MultiPath -> {
                val idx = result.recommendedIndex - 1
                if (idx in result.paths.indices) result.paths[idx].body else null
            }
        }
    }
}